File: sorting/quicksort/component/quicksort.rb
class Array
  def quicksort
    return self if self.length < 2
    less = Array.new
    gret = Array.new
    equal = Array.new
    pivot = (self.first + self.last) / 2
    self.each do |n|
      if n > pivot
        gret.push(n)
      elsif n == pivot
        equal.push(n)
      else
        less.push(n)
      end
    end
    gret = gret.quicksort
    less = less.quicksort
    return less + equal + gret
  end
end

File: sorting/quicksort/component/quicksort.js
const QuickSort = (array) => {
  const len = array.length;
  if (len < 2) return array;

  const pivot = array[Math.floor(len / 2)];
  const lesser = [];
  const greater = [];
  const equal = [];

  for (var i = 0; i < len; i++) {
    let num = array[i];
    if (num < pivot) {
      lesser.push(num);
    } else if (num > pivot) {
      greater.push(num);
    } else {
      equal.push(num);
    }
  }
  return [].concat(QuickSort(lesser), equal, QuickSort(greater));
};

export default QuickSort;
